core/arch/arm/plat-imx/config/config_imx7.h

```diff
@@ -21,9 +23,18 @@
 #define DRAM0_BASE		0x80000000
 #define DRAM0_SIZE		CFG_DDR_SIZE
 
-/* Location of trusted dram */
-#define TZDRAM_BASE		(DRAM0_BASE + CFG_DDR_SIZE - 32 * 1024 * 1024)
-#define TZDRAM_SIZE		(30 * 1024 * 1024)
+/*
+ * Location of trusted dram
+ * Env may provide CFG_DDR_TEETZ_RESERVED_START/_SIZE: TEE/TA_RAM + SHM.
+ * Otherwise, locate at DDR top (last 32MB minus SHM size).
+ */
+#ifdef CFG_DDR_TEETZ_RESERVED_START
```

Comment:
As for imx6multi, this introduce `CFG_DDR_TEETZ_RESERVED_START`/`_SIZE` as config options which were not previously. I will discard such change. Maintainers will be free to later add support for external config means (`CFG_DDR_TEETZ_RESERVED_START` or `CFG_TZDRAM_BASE`) if they will.

Reply:
This is the issue I ran into. I think `CFG_DDR_TEETZ_RESERVED_` has the wrong semantics. `CFG_DDR_TEETZ_RESERVED_` is the region of memory reserved for the TEE core, TA's, and the shared memory area. It assumes all regions are contiguous, and shared memory is located in this region. My understanding is that the shared memory region does not need to be locked down to trustzone, because normal world needs access to it. Therefore I think it is wrong to assume that the shared memory region is contiguous with TEE private memory.

I think it would be better to expose CFG_TZDRAM_BASE/SIZE and CFG_SHMEM_BASE/SIZE, to allow placement and sizing of TEE memory separately from shared memory. Another reason to retreat from `CFG_DDR_TEETZ_RESERVED_` is that it's not used in C code anywhere, so it's usage is purely based on convention. It could take on different meanings in different platforms, which would be confusing.

---
In reply to: [183714283](https://github.com/OP-TEE/optee_os/pull/2273#discussion_r183714283) [](ancestors = 183714283)

Reply:
Well, as maintainer of the stm that relies on `CFG_DDR_TEETZ_RESERVED_START/_SIZE`, I would like it to be maintained for my board. For the imx platform, I can't tell. No other platform really use `CFG_TZDRAM_BASE/_SIZE` setups.
If a generic `CFG_TZDRAM_BASE/_SIZE` comes, I will look at porting the stm paltform, but I don"t want to break it straight.
There are platforms that currently do not offer means to configure the memory layout . There is no need to force such to. Maybe the HW they use would not support the changes.

About behind the words:
* `CFG_DDR_TEETZ_RESERVED_START/_SIZE` covers TEE_RAM + TA_RAM + Static SHM. They form a nice couple to define the memory carved out from the Linux main memory.
* `CFG_TZDRAM_BASE` +`_SIZE` defines the Secure DRAM. So platform must provide also `CFG_SHMEM_START` + `_SIZE` for the static shm. Nice also. 


I would like `CFG_TZDRAM_*`  to cover also the emulated SRAM when enabling pager without a real SRAM to operate in. No need for 2 walues (with/without pager) and an added `EMUL_SRAM_START`.